File: Data Structures & Algorithms Module/Week6/LinkedStack.java
/*
LinkedStack.java
Billy Cussen
01/10/2020
*/

public class LinkedStack<T> implements StackInterface<T>{

    Node<T> first;

    @Override
    public void push(T e){
        if(isEmpty()){
            this.first = new Node<T>(e, null);
        } else{
            this.first = new Node<T>(e, this.first);
        }
    }
    
    @Override
    public T pop(){
        if(isEmpty()){
            throw new NullPointerException();
        } else{
            T element = this.first.element;
            first = this.first.next;
            return element;
        }
    }

    @Override
    public T peek(){
        return this.first.element;
    }

    @Override
    public int size(){
        return sizeHelper(this.first);
    }

    @Override
    public String toString(){
        String result = "***LINKED STACK TRAVERSAL***\n";
        int position = 0;
        Node<T> current = this.first;
        return toStringHelper(result,position,current);
    }

    private int sizeHelper(Node<T> current){
        if(current == null){
            return 0;
        }
        return 1 + sizeHelper(current.next);
    }

    private String toStringHelper(String result, int position, Node<T> current){
        if(current == null){
            return result;
        }
        result+="Position "+position+", Element: "+current.element+"\n";
        return toStringHelper(result,++position,current.next);
    }

    private boolean isEmpty(){
        return this.first == null;
    }

}

File: Python/Data Structures & Algorithms Module - Python/Week4/LinkedStack.py
"""
LinkedStack.py
Billy Cussen
12/11/2020
"""

from Node import Node

class LinkedStack:

    def __init__(self):
        self.first = None

    def isEmpty(self):
        return self.first == None

    def push(self, elem):
        if self.isEmpty():
            self.first = Node(elem, None)
        else:
            self.first = Node(elem, self.first)

    def pop(self):
        if self.isEmpty():
            raise Exception("Can't pop, the Stack is Empty!")
        elem = self.first.element
        self.first = self.first.next
        return elem

    def peek(self):
        if self.isEmpty():
            raise Exception("Can't peek, the Stack is Empty!")
        return self.first.element

    def mySize(self):
        return self.mySizeHelper(self.first)

    def mySizeHelper(self, current):
        if current == None:
            return 0
        return 1 + self.mySizeHelper(current.next)

    def toString(self):
        if self.isEmpty():
            raise Exception("Can't traverse, the Stack is Empty!")
        result = "***LINKED STRING TRAVERSAL***\n"
        position = 0
        return self.toStringHelper(self.first,result,position)

    def toStringHelper(self, current, result, position):
        if current == None:
            result+="***END***\n"
            return result
        result+="Position "+str(position)+", Element: "+str(current.element)+"\n"
        position+=1
        return self.toStringHelper(current.next,result,position)
        